### PR TITLE
MF-3: Allow for infrastructure scripts to be modified/configured in openmrs-module-spa

### PR DESCRIPTION
**Pull request by fatmali**

### CODE REVIEW

**Comment by fatmali** [comment]
PR addresses this JIRA ticket https://issues.openmrs.org/browse/MF-3

**Comment by dkayiwa** [comment]
Shouldn't the module bundle these files by default such that once i install the module, i have everything working out of the box?

**Comment by fatmali** [comment]
Hi @dkayiwa , I thought @ojwangantony already worked on this in a different PR. @joeldenning am I right?

**Comment by brandones** [comment]
@fatmali , @ojwangantony worked on a related PR, https://github.com/openmrs/openmrs-module-spa/pull/6 . This creates the `frontend` directory in the App Data Dir, which gets served at `/spaResources`. It doesn't, however, load these static systemjs resources into that directory.

I recommend that we make [SpaActivator:createFrontendDirectory](https://github.com/openmrs/openmrs-module-spa/blob/2aef8dfcc1f72755040d29d41940b615aa14c1e3/omod/src/main/java/org/openmrs/module/spa/SpaActivator.java#L46) copy those files (from `resources`, say) into the app data `frontend/` directory so that they will be available at `/openmrs/spaResources/systemjs@4.1.0/system.min.js` etc.

Just to be clear: I'm suggesting that these resources should not be made available under `frontend/` because this is the path for frontend routes -- every path that begins with `frontend/` should be expected to be handled by the frontend.

(wasn't sure whether to post this here or on the ticket -- this is somewhat in response to Joel's comments on the ticket as well)

**Comment by ojwanganto** [comment]
> @fatmali , @ojwangantony worked on a related PR, #6 . This creates the `frontend` directory in the App Data Dir, which gets served at `/spaResources`. It doesn't, however, load these static systemjs resources into that directory.
> 
> I recommend that we make [SpaActivator:createFrontendDirectory](https://github.com/openmrs/openmrs-module-spa/blob/2aef8dfcc1f72755040d29d41940b615aa14c1e3/omod/src/main/java/org/openmrs/module/spa/SpaActivator.java#L46) copy those files (from `resources`, say) into the app data `frontend/` directory so that they will be available at `/openmrs/spaResources/systemjs@4.1.0/system.min.js` etc.
> 
> Just to be clear: I'm suggesting that these resources should not be made available under `frontend/` because this is the path for frontend routes -- every path that begins with `frontend/` should be expected to be handled by the frontend.
> 
> (wasn't sure whether to post this here or on the ticket -- this is somewhat in response to Joel's comments on the ticket as well)

@brandones, the suggestion is great. That way, there is an assurance that the files are made available when the module is installed. Putting them in the frontend directory, maybe a in a sub-directory, also allows for updates to the files without the need to update the module and have it upgraded.

**Comment by fatmali** [comment]
Hi @brandones, thanks for your response.

> I recommend that we make [SpaActivator:createFrontendDirectory](https://github.com/openmrs/openmrs-module-spa/blob/2aef8dfcc1f72755040d29d41940b615aa14c1e3/omod/src/main/java/org/openmrs/module/spa/SpaActivator.java#L46)  copy those files (from resources, say) 

Is `resources` the agreed upon dir where the static files will be stored? is it configurable?

**Comment by brandones** [comment]
When I say `resources` I’m referring to the thing in the module, you know, `org.openmrs.module.spa.resources` or whatever it is. What I’m saying is that we should keep a copy of systemjs in there, just so that users never need to download it. SpaActivator should copy those files from `resources` into the SPA static files directory, which is called `frontend` and lives in the application data directory. This will cause them to be served via the mechanism that Antony implemented.

**Comment by internettrans** [comment]
> It doesn't, however, load these static systemjs resources into that directory.

Generating the correct frontend directory for those doing self-hosted js assets is one of the next main projects I've been wanting to work on with you, @brandones and @ojwangantony. Fatma's PR is not going to handle that, but rather a whole separate PR or even project that does this. The idea is that we will generate the correct frontend directory from a package.json file by doing an npm install. The `systemjs@4.1.0/dist/system.min.js` file will come from the `node_modules` directory.

> This creates the frontend directory in the App Data Dir, which gets served at /spaResources. 

Did we change it so that in the browser you have to go to `/spaResources`, not to `/frontend` to get the assets? I must have missed that change. If so, then yes @fatmali we need to change this PR to have `/spaResources` in it instead of `/frontend`.

> I recommend that we make SpaActivator:createFrontendDirectory copy those files (from resources, say) into the app data frontend/ directory so that they will be available at /openmrs/spaResources/systemjs@4.1.0/system.min.js etc.

See my previous comment -- that is out of scope of this PR and also something I don't think we should even do with java code. Instead, it should be done with an npm install and then moving some or all of the node_modules into the frontend directory.

> What I’m saying is that we should keep a copy of systemjs in there, just so that users never need to download it.

I disagree -- would prefer to do this with npm and a package.json instead of java inside of openmrs-module-spa.

**Comment by brandones** [comment]
> Did we change it so that in the browser you have to go to /spaResources, not to /frontend to get the assets? I must have missed that change. If so, then yes @fatmali we need to change this PR to have /spaResources in it instead of /frontend.

Yes, so as to make resources paths not collide with frontend paths.

> I disagree -- would prefer to do this with npm and a package.json instead of java inside of openmrs-module-spa.

Oh yeah, this is 100% the right approach.

**Comment by ojwangantony** [comment]
I don't think we changed from /frontend. It's the servlet that is accessed
through /spaResources. All frontend resources are therefore accessible
through /frontend

**Comment by internettrans** [comment]
> I don't think we changed from /frontend. It's the servlet that is accessed
through /spaResources. All frontend resources are therefore accessible
through /frontend

@ojwangantony can you help me understand this? Given the following HTTP requests, which ones serve the static frontend assets?

```sh
# Which request to the server serves up the frontend assets?

# Option 1 - /frontend
/frontend/hi.js

# Option 2 - /spaResources
/spaResources/hi.js
```

**Comment by ojwangantony** [comment]
It's /frontend/xx.js. There is internal rerouting to /spaResources by the
module servlet as can be seen in the code. Am not on my machine and can't
explain this better without referencing the code.

File: omod/src/main/webapp/master-single-page-application.jsp [hunk 1]
```diff
@@ -7,14 +7,14 @@
...
-    <script src="https://cdn.jsdelivr.net/npm/import-map-overrides@1.7.2/dist/import-map-overrides.min.js"></script>
-    <script src="https://cdnjs.cloudflare.com/ajax/libs/systemjs/4.1.0/system.min.js"></script>
-    <script src="https://cdnjs.cloudflare.com/ajax/libs/systemjs/4.1.0/extras/amd.min.js"></script>
-    <script src="https://cdnjs.cloudflare.com/ajax/libs/systemjs/4.1.0/extras/named-exports.js"></script>
-    <script src="https://cdnjs.cloudflare.com/ajax/libs/systemjs/4.1.0/extras/named-register.min.js"></script>
-    <script src="https://cdnjs.cloudflare.com/ajax/libs/systemjs/4.1.0/extras/use-default.min.js"></script>
+    <link rel="preload" href="${cookie['import-map-override-url'] == null ? '/frontend/import-map.json' : cookie['import-map-override-url'].getValue()}" as="fetch" crossorigin="anonymous" />
+    <script type='systemjs-importmap' src="${cookie['import-map-override-url'] == null ? '/frontend/import-map.json' : cookie['import-map-override-url'].getValue()}"></script>
+    <script src="/frontend/import-map-overrides@1.7.2/dist/import-map-overrides.min.js"></script>
+    <script src="/frontend/systemjs@4.1.0/system.min.js"></script>
```
Comment (internettrans) [inline omod/src/main/webapp/master-single-page-application.jsp:13]: The switch from `/` to `@` is a good one I think. https://www.jsdelivr.com/ is a good CDN to use for Ampath and others who want to reverse proxy these internal urls to external urls, and jsdelivr [uses @ in its urls](https://www.jsdelivr.com/package/npm/systemjs)

**Review by internettrans — APPROVED** [review]
👍 this looks good to me, since `/frontend` is the correct route.

@brandones @ojwanganto any further comments before we merge this?

**Comment by dkayiwa** [comment]
> I disagree -- would prefer to do this with npm and a package.json instead of java inside of openmrs-module-spa.

Will this work without an internet connection?

**Comment by internettrans** [comment]
@dkayiwa npm install does not work offline. However, the `npm install` would happen when packaging the distribution, not when executing the code. So yes, by the time the code is being executed, no internet connection will be required.

**Comment by dkayiwa** [comment]
@joeldenning just to make sure i understand this, will everything get bundled into a .omod as i build the module?

**Comment by internettrans** [comment]
No omod will be necessary, since it's just javascript files being put in the `frontend` or `spaResources` directory.

**Comment by dkayiwa** [comment]
I think what is confusing me is that this is an OpenMRS server side module which we distribute to the hospital as a .omod
For modules that have resources like those used in the reference application, they are bundled together with the varius .omods, on a CD or flash disk, such that on installation, they just run without requiring a hospital to have an internet connection.

I do not yet understand how this fits in.

**Comment by brandones** [comment]
> It's /frontend/xx.js. There is internal rerouting to /spaResources by the module servlet as can be seen in the code. Am not on my machine and can't explain this better without referencing the code.

I don’t understand this at all and would love it if someone would explain at the next meeting.

**Comment by brandones** [comment]
@dkayiwa Yes, the architecture for this project is more like Initializer or Bahmni configs. The implementer needs to figure out how to deploy these files into the application data directory on servers. At PIH, we use Puppet to install configuration files, and we would do the same for frontends. If you’re stuck deploying only omods, perhaps you should package the files you need into an omod that just copies its resources directory into the app data directory. But it will probably be worthwhile to figure out a deployment mechanism for other kinds of artifacts. I think keeping configuration in modules is something everyone should be moving away from.

**Review by brandones — APPROVED** [review]

**Comment by internettrans** [comment]
> I think what is confusing me is that this is an OpenMRS server side module which we distribute to the hospital as a .omod
> For modules that have resources like those used in the reference application, they are bundled together with the varius .omods, on a CD or flash disk, such that on installation, they just run without requiring a hospital to have an internet connection.
> 
> I do not yet understand how this fits in.

Yeah it's a new thing 😄. Bundling javascript code into a jar/omod via Maven is something I intentionally chose to avoid, since the frontend world is capable of packaging its own code and it would be a pain to get Maven/Java to do what node/npm were built to do.

There are two situations to consider:

**Connectivity required at runtime** (ampath, openmrs-spa.org, maybe bahmni):

For these situations, the application data frontend directory will not be used at all for anything. Instead, a load balancer such as nginx is used to route requests from `/frontend` to a CDN such as AWS S3, Digital Ocean Spaces, etc. In other words, the js assets are not even served by Java/Tomcat at all. This is how openmrs-spa.org and ampath's application are working.

**No connectivity required at runtime** (majority of distributions):

The application data frontend directory must be populated by the developer/implementor when they're creating the WAR file to ship off to an offline location. Brandon, Antony, and I will soon be creating a repo and implementing a tool that facilitates this. The tool will not be written in java, but rather in node. It will perform an `npm install` and then run `npm run build`. The `npm run build` command is sort of the nodejs equivalent of `mvn clean install`. The `npm run build` will produce a `dist` directory that has an import map json file and all other static js/html/img assets in it. That dist directory will need to be copied into the Application Data's frontend directory before everything is zipped/war'ed up to go to the offline location.

**Comment by internettrans** [comment]
A third option I didn’t list above ^ is situations where the implementor chooses to put nginx or similar in front of tomcat, and then routes the requests to /frontend to a self hosted version of the assets. This does not require runtime connectivity but would still use the tool I mentioned above that packages the frontend assets via node and npm.

The difference doesn’t actually relate to connectivity as much as whether a load balancer is present or not. It just so happens that the groups who have pushed for connectivity requirements have also been sophisticated in their load balancing.

**Pull request merged**